### DISCUSSION TITLE
Stopping indentation with selected text when Tab is pressed

### DISCUSSION THREAD

**Discussion by kofa73** [atom/atom#22691] [Q&A]
Hello,

In most of my editors and IDEs, if I select a piece of text and type something, the selected text is deleted on the first keystroke, and is replaced by the text I type. However, pressing **Tab** is, in some editors, special: it will indent the text (as if Tab were pressed with the cursor at the beginning of the line). Atom is one such editor.

When editing tab-separated text files, I'd really like the text to be overwritten: I want to enter a Tab (column separator) where I am, overwriting any selected text, instead of at the beginning of the line. However, I have not found a preference to allow this. Is this possible to achieve somehow?

Thanks,
Kofa

**Comment by ferdnyc** [comment]
@kofa73 

A slightly tricky one, since it requires overriding a built-in keybinding. And since you want to replace the selected text with something different, it also requires a custom command. (Or seems to, that's how I solved it anyway.)

It doesn't help that the documentation for custom commands in the "Keymaps In-Depth" section of the Flight Manual is out of date, and shows an example custom command that won't work since it's written in JavaScript, not CoffeeScript.

Here's how I solved it:
1. Edit your `init.coffee` file, easiest way is Edit > Init Script... from the menu.
1. Add the following custom command:
   ```coffee
    atom.commands.add 'atom-text-editor', 'custom:delete-and-tab', ->
      return unless editor = atom.workspace.getActiveTextEditor()
      selection = editor.getLastSelection()
      editor.cutSelectedText()
      selection.insertText "\t"
    ```
1. Edit your keybindings, again easiest way is Edit > Keymaps... from the menu
1. Add the following <s>YAML-I-think?</s>CoffeeScript Object Notation binding:
    ```coffee
    'atom-text-editor:not([mini])':
      'tab': 'custom:delete-and-tab'
    ```
1. Restart Atom

That should override the default mapping for `tab` (which is `editor:indent`) in the `atom-text-editor:not([mini])` context, without altering other meanings of `tab` in other contexts. Anywhere `tab` doesn't have a special meaning, typing `tab` will always replace the selection (if any) with a single hard `tab` character.

It will most likely override the default meaning of `tab` in more places than you'd really like, but it's possible to restrict the new mapping to certain file types, or certain contexts... see the docs for more info. Hopefully this at least gets you started.

There may be issues with brute-forcing things quite this hard that I haven't spotted, caveat customizer.

**Reply by ferdnyc** [reply]
The custom command as written will actually put the text removed on your clipboard... which could be handy, actually. However, if that's not what you want, just remove the `editor.cutSelectedText()` line. Without that, it'll still overwrite the contents of the current selection with the tab character. (Undo can still reverse that, of course.)

**Reply by ferdnyc** [reply]
(Last note, I swear...)

With the updated binding in place for <kbd>Tab</kbd>, you can still use <kbd>Ctrl</kbd>+<kbd>]</kbd> and <kbd>Ctrl</kbd>+<kbd>[</kbd> to indent and outdent, respectively, the selected-or-current line(s) of text.

I actually prefer those combos. Probably why I never even realized <kbd>Tab</kbd> is bound the way it is.

**Reply by kofa73** [reply]
Thanks for the detailed description.
